src/Radicle/Internal/Effects/Capabilities.hs

```diff
@@ -57,23 +65,31 @@ class (Monad m) => System m where
       -> m (Maybe Handle, Maybe Handle, Maybe Handle, ProcessHandle)
     waitForProcessS :: ProcessHandle -> m ExitCode
     hPutStrS :: Handle -> Text -> m ()
-    hGetLineS :: Handle -> m Text
+    -- | hGetLine should return Nothing on EOF.
+    hGetLineS :: Handle -> m (Maybe Text)
+    hCloseS :: Handle -> m ()
 
 instance System m => System (Lang m) where
     systemS proc = lift $ systemS proc
     waitForProcessS = lift . waitForProcessS
     hPutStrS a b = lift $ hPutStrS a b
     hGetLineS = lift . hGetLineS
+    hCloseS = lift . hCloseS
 instance System IO where
     systemS = createProcess
     waitForProcessS = waitForProcess
-    hPutStrS = hPutStr
-    hGetLineS x = toS <$> hGetLine x
+    hPutStrS h t = hSetBuffering h LineBuffering >> hPutStr h t >> hFlush h
+    hGetLineS x =
+        catchJust (\e -> if isEOFError e then Just () else Nothing)
+                  (Just . toS <$> (hSetBuffering x LineBuffering >> hGetLine x))
+                  (\() -> pure Nothing)
```

Comment:
You could use [`hIsEOF`](http://hackage.haskell.org/package/base-4.12.0.0/docs/System-IO.html#v:hIsEOF) instead of `catchJust`

Reply:
Done.